packages/@sanity/schema/src/sanity/createSchemaFromManifestTypes.ts

```diff
@@ -1,25 +1,292 @@
+import {
+  type CustomValidator,
+  type MediaValidator,
+  type Rule as IRule,
+  type RuleTypeConstraint,
+  type UriValidationOptions,
+} from '@sanity/types'
+import {cloneDeep, isObject} from 'lodash'
+
 import {Schema} from '../legacy/Schema'
 import {builtinTypes} from './builtinTypes'
+import {groupProblems} from './groupProblems'
+import {type SchemaValidationResult} from './typedefs'
+import {validateSchema} from './validateSchema'
+import {ValidationError} from './validation/ValidationError'
 
 const builtinSchema = Schema.compile({
   name: 'studio',
   types: builtinTypes,
 })
 
 export function createSchemaFromManifestTypes(schemaDef: {name: string; types: unknown[]}) {
+  const validated = validateSchema(schemaDef.types).getTypes()
+  const validation = groupProblems(validated)
+  const problems = validation.filter((group) =>
+    group.problems.some((problem: SchemaValidationResult) => problem.severity === 'error'),
+  )
+
+  if (problems.length > 0) {
+    throw new ValidationError(problems)
+  }
+
   return Schema.compile({
     name: schemaDef.name,
     types: schemaDef.types.map(coerceType).filter(Boolean),
     parent: builtinSchema,
   })
 }
 
-function isRecord(val: unknown): val is Record<string, unknown> {
-  return val !== null && typeof val === 'object' && !Array.isArray(val)
+// coerceType attempts to coerce a json schema into a Sanity schema type. This mainly involves
+// converting a json normalized rule into an actual '@sanity/schema' Rule.
+function coerceType(obj: unknown) {
+  if (!isObject(obj)) return undefined
+  const typ = cloneDeep(obj)
+  traverse(typ)
+  return typ
+}
+
+function traverse(obj: unknown) {
+  if (!isObject(obj)) {
+    return
+  }
+
+  if (Array.isArray(obj)) {
+    obj.forEach(traverse)
+    return
+  }
+
+  for (const v of Object.values(obj)) {
+    traverse(v)
+  }
+
+  coerceValidation(obj)
+}
+
+// Convert a json rule into a '@sanity/schema' Rule.
+function coerceValidation(val: unknown) {
+  if (!isObject(val) || !('validation' in val)) return
+
+  // Convert ManifestValidationGroup[] to Rule functions
+  const manifestValidation = Array.isArray(val.validation) ? val.validation : [val.validation]
+
+  val.validation = manifestValidation
+    .map((group) => {
+      if (!isObject(group)) return undefined
+
+      return (baseRule: IRule) => {
+        let rule = baseRule
+        const level = 'level' in group ? group.level : undefined
```

Comment:
Nit: Simpler to just do:
```suggestion
        const level = group.level
```
?